Optim foreach cleanup for Adamax

mikaylagawarecki opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70953
* #70484
* #70483
* #70482
* #70481
* #70295
* #70231
* #70230
* #70229
* __->__ #69982
* #69981
* #69980
* #71335
* #71334
* #71333

Add foreach flag to Adamax optimizer + cleanup

Differential Revision: [D33767865](https://our.internmc.facebook.com/intern/diff/D33767865)

mikaylagawarecki committed:
Optim foreach cleanup for Adamax

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adamax"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adamax"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adamax"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adamax"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adamax"




[ghstack-poisoned]

mikaylagawarecki commented on torch/optim/adamax.py:
```diff
@@ -107,15 +111,120 @@ def step(self, closure=None):
+                         grads: List[Tensor],
+                         exp_avgs: List[Tensor],
+                         exp_infs: List[Tensor],
+                         state_steps: List[int],
```
The original (adamax)[https://github.com/pytorch/pytorch/blob/master/torch/optim/_multi_tensor/_functional.py#L73] in `optim/_multi_tensor/functional.py` takes in `state: List[Dict]` instead, but `states` was not used other than to access `state['steps']` for each `state` in `states` so I changed this to take in `state_steps` so the inputs would be the same as the single tensor version, is this ok?

albanD replied:
I don't think we have to take exactly the same thing since these are not public and we have a function above that can massage the arguments. But it is nicer if we can for sure.

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adamax"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adamax"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adamax"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adamax"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adamax"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adamax"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adamax"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adamax"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adamax"




[ghstack-poisoned]

albanD commented on torch/distributed/optim/functional_adamax.py:
```diff
@@ -100,4 +102,5 @@ def step(self, gradients: List[Optional[Tensor]]):
                      lr=self.defaults['lr'],
-                     weight_decay=self.defaults['weight_decay'])
+                     weight_decay=self.defaults['weight_decay'],
+                     foreach=self.foreach)
```
cc @mrshenli are these objects ever serialized? Should we take extra care to make sure an old saved version still works?

albanD reviewed (commented):
SGTM

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adamax"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adamax"




[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33767865).

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adamax"


Differential Revision: [D33767865](https://our.internmc.facebook.com/intern/diff/D33767865)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33767865).
